// jwt-validate@0.5.0/dist/esm/utils.mjs
export var CloudType;
(function (CloudType) {
    CloudType[CloudType["Public"] = 0] = "Public";
    CloudType[CloudType["Ppe"] = 1] = "Ppe";
    CloudType[CloudType["USGovernment"] = 2] = "USGovernment";
    CloudType[CloudType["China"] = 3] = "China";
})(CloudType || (CloudType = {}));
/**
 * Retrieves the JWKS URI for the specified tenant.
 * @param {string} [tenant='common'] - The tenant to retrieve the JWKS URI for.
 * @param {CloudType} [cloud=CloudType.Public] - The cloud to retrieve the JWKS URI for.
 * @returns {Promise<string>} - A promise that resolves with the JWKS URI.
 */
export async function getEntraJwksUri(tenant = 'common', cloud = CloudType.Public) {
    let cloudUrl = '';
    switch (cloud) {
        case CloudType.Public:
            cloudUrl = 'login.microsoftonline.com';
            break;
        case CloudType.Ppe:
            cloudUrl = 'login.windows-ppe.net';
            break;
        case CloudType.USGovernment:
            cloudUrl = 'login.microsoftonline.us';
            break;
        case CloudType.China:
            cloudUrl = 'login.chinacloudapi.cn';
            break;
    }
    const res = await fetch(`https://${cloudUrl}/${tenant}/.well-known/openid-configuration`);
    const data = (await res.json());
    return data.jwks_uri;
}
